model_compression_toolkit/data_generation/keras/optimization_utils.py

```diff
@@ -0,0 +1,444 @@
+            bn_layer_mean = self.orig_bn_stats_holder.get_mean(bn_layer_name=layer_name)
+            bn_layer_var = self.orig_bn_stats_holder.get_var(bn_layer_name=layer_name)
+
+            # Get the mean and variance from the current batch's statistics
+            if self.use_all_data_stats:
+                # If using all data statistics, retrieve the accumulated statistics from all batches
+                imgs_layer_mean, imgs_layer_var = self.get_accumulated_stats_per_layer(layer_name=layer_name)
+            else:
+                # Otherwise, retrieve the statistics from the current batch
+                imgs_layer_mean, imgs_layer_var = (
+                    self.all_imgs_stats_holder.get_stats(batch_stats_holder=batch_stats_holder, layer_name=layer_name))
+
+            bn_layer_std = tf.sqrt(bn_layer_var + self.eps)
+            imgs_layer_std = tf.sqrt(imgs_layer_var + self.eps)
+
+            total_bn_loss += layer_weight * self.bn_loss_fn(bn_mean=bn_layer_mean,
+                                                            input_mean=imgs_layer_mean,
+                                                            bn_std=bn_layer_std,
+                                                            input_std=imgs_layer_std)
+        return total_bn_loss
+
+    def compute_output_loss(self,
+                            output_imgs: tf.Tensor,
+                            activation_extractor: KerasActivationExtractor,
+                            tape: tf.GradientTape
+                            ) -> tf.Tensor:
+        # If output_loss_multiplier is zero return 0
+        return self.output_loss_multiplier * self.output_loss_fn(
+            output_imgs=output_imgs,
+            activation_extractor=activation_extractor,
+            tape=tape,
+            weights_last_layer=
+            self.get_model_output_weight_norm(model=self.model, activation_extractor=activation_extractor)) \
+            if self.output_loss_multiplier > 0 else tf.zeros(1)
+
+    def get_model_output_weight_norm(self,
+                                     model: tf.keras.Model,
+                                     activation_extractor: KerasActivationExtractor) -> Optional[tf.Tensor]:
+        """
+        Get the weight tensor of the last Dense or Conv2D layer in the model.
+
+        Args:
+            model (Model): The model to retrieve the weight tensor from.
+            activation_extractor (ActivationExtractor): Extractor for layer activations.
+
+        Returns:
+            Optional[tf.Tensor]: The weight tensor of the last linear layer, or None if not found.
+        """
+        # Iterate through the layers in reverse order.
+        for layer in reversed(model.layers):
+            if isinstance(layer, activation_extractor.linear_layers):
+                # Return the weight tensor of the layer.
+                return layer.weights[0] if layer.weights else None
+
+        # If no suitable layer is found, return None.
+        return None
```

Comment:
I couldn't find where is the return value of this method being used, but need to make sure this method knows to handle a case where None is returned